Edited Aevol/doc/aevol_run.1
./"test with man -l <file>
.TH AEVOL "1" "July 2014" "aevol 4.4" "User Manual"
.SH NAME
aevol \- an in silico experimental evolution platform
.SH SYNOPSIS
.B aevol_run \-h
|
.B \-\-help
.br
.B aevol_run \-V
|
.B \-\-version
.br
.B aevol_run \fR[\fB\-r\fI GENER\fR] \fR[\fB\-n\fI NB_GENER\fR] \fR[\fB\-tvwx\fR]
.SH DESCRIPTION
.B Aevol
is a simulation platform that allows one to let populations of digital organisms evolve in different conditions and study experimentally the mechanisms responsible for the structuration of the genome and the transcriptome.
.TP
.B aevol_run
run an aevol simulation.
.SH OPTIONS
.TP
.B \-h, \-\-help
print help, then exit
.TP
.B \-V, \-\-version
print version number, then exit
.HP
.B \-r, \-\-resume
.I  GENER
.br
specify generation to resume simulation at (default 0)
.HP
.B \-n, \-\-nbgener
.I NB_GENER
.br
specify number of generations to be run (default 1000)
.TP
.B \-t, \-\-text
use text files instead of binary files when possible
.TP
.B \-v, \-\-verbose
be verbose
.TP
.B \-w, \-\-wait
pause after loading
.TP
.B \-x, \-\-noX
don't display X outputs upon start
.br
send SIGUSR1 to switch X output on/off
.SH "SEE ALSO"
.B aevol_create, aevol_modify, aevol_propagate


Edited Aevol/ChangeLog
New in 8.0:
 - MPI Support:
    - Fully reproduceable MPI execution with sequential and memory shared parallel version (OpenMP)
    - Can be mixed up with OpenMP version: MPI+OpenMP
    - No (Explicit) Synchronization (i.e. no MPI Barrier). Should be very scalable.
    - RAevol (dynamic and static environment) and Aevol compatible
    - Lineage/Tree are fully supported
    
 - RAevol should be fixed
 - A lot of vectorization works on motif searchs function
 - Validated discrete fuzzy (against the continious fuzzy). WARNING : never compare results of discrete and continious fuzzy
 - A lot of memory leaks fixes
 - A lot of bug fixes
 - New differential search algorithm that does the diff for all metadatas (not only promoters all the metadatas Promoters/RNAs/Genes)

New in 4.4:

* aevol_create now gives a name to the created strain. It can be defined using
  the STRAIN_NAME keyword in the param file. If this keyword is not present, a
  default one will be generated.

* New option -o or --out in aevol_create
  Allows one to create the experiment in any location.

* aevol_propagate can now reset the PRNGs (random generators) with the new
  general option -S (affecting all random processes) or the new more specific
  options -m, -s, -t, -e and -n (each affects a single random process, see man
  page).

* The environmental target can be defined by a list of custom points using the
  ENV_ADD_POINT keyword in the parameter file. This is an alternative to the
  usual way, where it is defined as a sum of gaussians. Note, however, that
  environmental variation is possible only for the "gaussian" way.

* The aevol_misc_fixed_mutations program now outputs the number of genes
  affected by each mutation. More precisely, three new columns are produced:
    - the number of coding RNAs possibly disrupted by a switch, a small indel
      or a rearrangement breakpoint,
    - the number of coding RNAs entirely included in a rearranged segment,
    - in the case of a transfer by replacement, the number of genes that were
      entirely included in the replaced segment.
  This program does not work yet if plasmids are allowed.

* The programs aevol_misc_lineage, aevol_misc_ancstats,
  aevol_misc_fixed_mutations, aevol_misc_gene_families now work when the
  environment was changed at some point by aevol_modify.

* The fitness proportionate selection scheme now works with spatial structure

* In aevol_modify, it is now possible to replace a population by clones of the
  best individual. To do so, add the line CLONE_BEST_NO_TREE or
  CLONE_BEST_AND_CHANGE_TREE in the parameter file you supply to aevol_modify.
  Use CLONE_BEST_AND_CHANGE_TREE if your simulation campaign uses the
  genealogical trees, that is to say if you set RECORD_TREE to true. Otherwise,
  use CLONE_BEST_NO_TREE.

* Manual pages were added for aevol_misc_lineage, aevol_misc_ancstats,
  aevol_misc_fixed_mutations, aevol_misc_gene_families, aevol_misc_create_eps,
  aevol_misc_view_generation.

* New option (-c and -p) in aevol_create. Allows one to load a chromosome (and
  possibly plasmid) from a text file. If -p is used, ALLOW_PLASMIDS must be set
  to true and -c must also be used. If -c is used and ALLOW_PLASMIDS is set to
  true, -p must also be used.

* Add a script (src/misc/movies.py) to produce a movie from an aevol simulation
  with dumps activated.
  run ./src/misc/movies.py -h for help.
  Needs ffmpeg in path.

* The aevol_misc_gene_families program issues the detailed history of each gene
  family on the lineage of a given individual (providing its lineage file).
  A gene family is defined here as a set of coding sequences that arised by
  duplications of a single original gene. The original gene, called the root of
  the family, can either be one of the genes in the initial ancestor, or a new
  gene created from scratch (for example by a local mutation that transformed a
  non-coding RNA into a coding RNA). The history of gene duplications, gene
  losses and gene mutations in each gene family is represented by a binary tree.
  The program starts by loading the initial genome at the beginning of the
  lineage and by tagging each gene in this initial genome. Each of these initial
  genes is marked as the root of a gene family. Then, each mutation recorded in
  the lineage file is replayed and the fate of all tagged genes is followed and
  recorded in their respective families. When a gene is duplicated, the
  corresponding node in one of the gene trees becomes an internal node, and two
  children nodes are added to it, representing the two gene copies. When a gene
  sequence is modified, the mutation is recorded in its corresponding node in
  one of the gene trees. When a gene is lost, the corresponding node in one of
  the gene trees is labelled as lost. When a new gene appears from scratch, i.e.
  not by gene duplication, it becomes the root of a new gene tree. Environmental
  variations are also replayed exactly as they occured during the main run.
  When all mutations have been replayed, several output files are written in a
  directory called gene_trees. Two general text files are produced. The file
  called gene_tree_statistics.txt contains general data on each gene family,
  like its creation date, its extinction date, or how many nodes it contained.
  The file called nodeattr_tabular.txt contains information about each node of
  each gene tree, like when it was duplicated or lost or how many mutations
  occurred on its branch. In addition, for each gene tree, two text files are
  generated: a file called genetree******-topology.tre contains the topology of
  the gene tree in the Newick format, and a file called
  genetree******-nodeattr.txt that contains the detailed list of events that
  happened to each node in the tree file, before it was either duplicated or
  lost.

  Usage: ae_misc_gene_families [-c | -n] [-t tolerance] -f lineage_file


Changes in 4.4:


Edited Aevol/src/aevol_run.cpp
// ****************************************************************************
//
//          Aevol - An in silico experimental evolution platform
//
// ****************************************************************************
//
// Copyright: See the AUTHORS file provided with the package or <www.aevol.fr>
// Web: http://www.aevol.fr/
// E-mail: See <http://www.aevol.fr/contact/>
// Original Authors : Guillaume Beslon, Carole Knibbe, David Parsons
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//
// ****************************************************************************

// ============================================================================
//                                   Includes
// ============================================================================
#include <cerrno>
#include <cinttypes>
#include <cstdlib>
#include <cstdio>
#include <csignal>

#include <sys/stat.h>
#include <getopt.h>
#include <stdlib.h>
#include <stdio.h>
#include <signal.h>
#include <sys/stat.h>

#ifdef _OPENMP
#ifndef __OPENMP_GPU
#include <omp.h>
#endif
#else
typedef int omp_int_t;
inline omp_int_t omp_get_thread_num() { return 0;}
inline omp_int_t omp_get_max_threads() { return 1;}
#endif

#ifdef HAVE_MPI
#include <mpi.h>
#endif

// =================================================================
//                            Project Files
// =================================================================
#ifdef __X11
  #include "ExpManager_X11.h"
#else
  #include "ExpManager.h"
#endif

#include "ExpManager_7.h"
#include "macros.h"

#include "aevol_version.h"

using namespace aevol;

#ifndef __NO_X
  void catch_usr1(int sig_num);
#endif

// Helper functions
void print_help(char* prog_path);
void interpret_cmd_line_options(int argc, char* argv[]);

// Command-line option variables
// static bool pause_on_startup = false;
static bool verbose = false;
static int64_t t0 = -1;
static int64_t t_end = -1;
static int64_t nb_steps = -1;
static int grain_size = 1;
static bool w_mrca = false;

#ifndef __NO_X
  static bool show_display_on_startup = true;
#endif

#ifdef _OPENMP
static bool run_in_parallel = false;
#endif

// Other file-scope variables
static ExpManager* exp_manager = nullptr;


int main(int argc, char* argv[]) {
  // Set handlers for signals to be caught
  #ifndef __NO_X
    signal(SIGUSR1, catch_usr1);
  #endif

  ExpManager_7::standalone_simd = true;
  
  interpret_cmd_line_options(argc, argv);

  // Print the hash and date of the commit used to compile this executable
  printf("Running aevol version %s\n", version_string);

  #ifdef HAVE_MPI
  printf("Initializing MPI Environment\n");
  MPI_Init(NULL, NULL);

  int world_rank;
  MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
  int world_size;
  MPI_Comm_size(MPI_COMM_WORLD, &world_size);
  #endif

  // =================================================================
  //                          Load the simulation
  // =================================================================
  #ifndef __NO_X
    exp_manager = new ExpManager_X11();
  #else
    exp_manager = new ExpManager();
  #endif

  exp_manager->load(t0, verbose, true
  #ifdef HAVE_MPI
  ,world_rank
  #endif
  );
  exp_manager->set_t_end(t_end);

  #ifdef HAVE_MPI
  if (world_size != exp_manager->exp_s()->nb_rank()) {
    printf("MANDATORY : the nb_rank specify through aevol_create should be the same than the one for aevol_run\n");
    exit(-1);
  }
  #endif


  // Make a numbered copy of each static input file
  // (dynamic files are saved elsewhere)
  // TODO (?)

  #ifndef __NO_X
    if (show_display_on_startup) {
      ((ExpManager_X11*) exp_manager)->toggle_display_on_off();
    }
  #endif

  // =================================================================
  //                         Run the simulation
  // =================================================================
  exp_manager->run_evolution();
  
  #ifdef HAVE_MPI
  MPI_Finalize();
  #endif

  delete exp_manager;
  //~ return EXIT_SUCCESS;
}


#ifndef __NO_X
  void catch_usr1(int sig_num) {
    signal(SIGUSR1, catch_usr1);

    printf("display on/off\n");
    if (exp_manager != nullptr) {
      ((ExpManager_X11*) exp_manager)->toggle_display_on_off();
    }
  }
#endif


/*!
  \brief

*/
void print_help(char* prog_path) {
  // Get the program file-name in prog_name (strip prog_path of the path)
  char* prog_name; // No new, it will point to somewhere inside prog_path
  if ((prog_name = strrchr(prog_path, '/'))) {
    prog_name++;
  }
  else {
    prog_name = prog_path;
  }

	printf("******************************************************************************\n");
	printf("*                                                                            *\n");
	printf("*                        aevol - Artificial Evolution                        *\n");
	printf("*                                                                            *\n");
	printf("* Aevol is a simulation platform that allows one to let populations of       *\n");
  printf("* digital organisms evolve in different conditions and study experimentally  *\n");
  printf("* the mechanisms responsible for the structuration of the genome and the     *\n");
  printf("* transcriptome.                                                             *\n");
	printf("*                                                                            *\n");
	printf("******************************************************************************\n");
  printf("\n");
	printf("%s: run an aevol simulation.\n", prog_name);
  printf("\n");
	printf("Usage : %s -h or --help\n", prog_name);
	printf("   or : %s -V or --version\n", prog_name);
	printf("   or : %s [-b TIMESTEP] [-e TIMESTEP|-n NB_TIMESTEPS] [-p NB_THREADS] [-vwx]\n",
         prog_name);
	printf("\nOptions\n");
	printf("  -h, --help\n\tprint this help, then exit\n");
	printf("  -V, --version\n\tprint version number, then exit\n");
  printf("  -b, --begin TIMESTEP\n");
  printf("\tspecify time t0 to resume simulation at (default read in last_gener.txt)\n");
  printf("  -e, --end TIMESTEP\n");
  printf("\tspecify time of the end of the simulation\n");
  printf("  -n, --nb-timesteps NB_TIMESTEPS\n");
  printf("\tspecify number of timesteps to be simulated (default 1000)\n");
  printf("  -p, --parallel NB_THREADS\n");
  printf("\trun on NB_THREADS threads (use -1 for system default)\n");
  printf("  -v, --verbose\n\tbe verbose\n");
  printf("  -w, --wait\n\tpause after loading\n");
  printf("  -x, --noX\n\tdon't display X outputs upon start\n");
  printf("\tsend SIGUSR1 to switch X output on/off\n");
}

void interpret_cmd_line_options(int argc, char* argv[]) {
  // Define allowed options
  const char* options_list = "hVb:e:n:vwxp:6:";
  static struct option long_options_list[] = {
      {"help",          no_argument,       nullptr, 'h'},
      {"version",       no_argument,       nullptr, 'V'},
      {"begin",         required_argument, nullptr, 'b'},
      {"end",           required_argument, nullptr, 'e'},
      {"nb-timesteps",  required_argument, nullptr, 'n'},
      {"verbose",       no_argument,       nullptr, 'v'},
      {"wait",          no_argument,       nullptr, 'w'},
      {"noX",           no_argument,       nullptr, 'x'},
      {"parallel",      required_argument, nullptr, 'p'},
      {"aevol_6",       no_argument,       nullptr, '6'},
      {0, 0, 0, 0}
  };

  // Get actual values of the CLI options
  int option;
  while ((option =
              getopt_long(argc, argv, options_list, long_options_list, nullptr))
         != -1) {
    switch (option) {
      case 'h' : {
        print_help(argv[0]);
        exit(EXIT_SUCCESS);
      }
      case 'V' : {
        Utils::PrintAevolVersion();
        exit(EXIT_SUCCESS);
      }
      case 'b' : {
        t0 = atol(optarg);
        break;
      }
      case 'e' : {
        if (nb_steps != -1) {
          Utils::ExitWithUsrMsg("use either option -n or -e, not both");
        }

        t_end = atol(optarg);
        w_mrca = true;
        break;
      }
      case 'n' : {
        if (t_end != -1) {
          Utils::ExitWithUsrMsg("use either option -n or -e, not both");
        }

        nb_steps = atol(optarg);
        break;
      }
      case 'v' : {
        verbose = true;
        break;
      }
      case 'w' : {
        // pause_on_startup = true;
        break;
      }
    case '6' : {
      ExpManager_7::standalone_simd = false;
      break;
    }
      case 'x' : {
        #ifdef __NO_X
        printf("%s: error: Program was compiled with __NO_X option, "
                 "no visualisation available.\n", argv[0]);
          exit(EXIT_FAILURE);
        #else
        show_display_on_startup = false;
        #endif

        break;
      }
      case 'p' : {
        #ifdef _OPENMP
        run_in_parallel = true;
          if (atoi(optarg) > 0) {
            omp_set_num_threads(atoi(optarg));
          }
        #endif
        break;
      }
      case 'g' : {
        grain_size = atoi(optarg);
        break;
      }
      default : {
        // An error message is printed in getopt_long, we just need to exit
        exit(EXIT_FAILURE);
      }
    }
  }

  // If t0 wasn't provided, use default
  if (t0 < 0) {
    t0 = OutputManager::last_gener();
  }

  // If t_end_ wasn't provided, set it according to nb_steps or use default
  // (run for 1000 timesteps)
  if (t_end < 0) {
    if (nb_steps >= 0) {
      t_end = t0 + nb_steps;
    }
    else {
      t_end = t0 + 1000;
    }
  }

  // It the user didn't ask for a parallel run, set number of threads to 1
  #ifdef _OPENMP
  if (not run_in_parallel) {
      omp_set_num_threads(1);
    }
  #endif
}
